server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1183,145 +1197,155 @@ private void writeUpdatedShardMetaDataAndComputeDeletes(
-                                final Set<String> originalShardBlobs = shardContainer.listBlobs(OperationPurpose.SNAPSHOT).keySet();
-                                final BlobStoreIndexShardSnapshots blobStoreIndexShardSnapshots;
-                                final long newGen;
-                                if (useShardGenerations) {
-                                    newGen = -1L;
-                                    blobStoreIndexShardSnapshots = buildBlobStoreIndexShardSnapshots(
-                                        originalShardBlobs,
-                                        shardContainer,
-                                        originalRepositoryData.shardGenerations().getShardGen(indexId, shardId)
-                                    ).v1();
-                                } else {
-                                    Tuple<BlobStoreIndexShardSnapshots, Long> tuple = buildBlobStoreIndexShardSnapshots(
-                                        originalShardBlobs,
-                                        shardContainer
-                                    );
-                                    newGen = tuple.v2() + 1;
-                                    blobStoreIndexShardSnapshots = tuple.v1();
-                                }
-                                allShardsListener.onResponse(
-                                    deleteFromShardSnapshotMeta(
-                                        survivingSnapshots,
-                                        indexId,
-                                        shardId,
-                                        snapshotIds,
-                                        shardContainer,
-                                        originalShardBlobs,
-                                        blobStoreIndexShardSnapshots,
-                                        newGen
-                                    )
-                                );
-                            }
-
-                            @Override
-                            public void onFailure(Exception ex) {
-                                logger.warn(
-                                    () -> format(
-                                        "%s failed to delete shard data for shard [%s][%s]",
-                                        snapshotIds,
-                                        indexId.getName(),
-                                        shardId
-                                    ),
-                                    ex
-                                );
-                                // Just passing null here to count down the listener instead of failing it, the stale data left behind
-                                // here will be retried in the next delete or repository cleanup
-                                allShardsListener.onResponse(null);
-                            }
-                        });
+                    for (int shardId = 0; shardId < shardCount; shardId++) {
```

Comment:
I cannot comment on the existing code. So I'll just comment it here:

The `shardCount` is computed with 
```
counts.stream().mapToInt(i -> i).max().orElse(0);
```
and `counts` is from the earlier 
```
INDEX_METADATA_FORMAT.read(metadata.name(), indexContainer, indexMetaGeneration, namedXContentRegistry)
                                .getNumberOfShards();
```

The above is inside a try-catch block that will return `null` on exception. The intention, IIUC, is to just log and continue processing. However, I think we will get a NPE when computing `shardCount` with stream if there is a `null` in the stream?

Since this is an existing behaviour, we can have a separate follow-up if you also think it's worth fixing.

Reply:
This is all going to go away soon, but no a `GroupedActionListener` uses `org.elasticsearch.common.util.concurrent.AtomicArray#asList` to compute the response it sends to its delegate, and this filters out `null` values. Deeply confusing indeed.

Reply:
TIL